### ISSUE TITLE
Port boringssl ML-KEM-768 implementation to openssl

### ISSUE THREAD

**Issue by nhorman** [openssl/project#1020]
Add native code to support ML-KEM-768.

Repurpose the C++ code from boringSsl

Note test suites will be added as part of https://github.com/openssl/project/issues/1013 and functional correctness asserted as part of openssl/project#997 

Split out openssl/project#1014 to make this task more manageable

**Comment by baentsch** [comment]
The concrete activities are
- Change the C++ code in https://boringssl.googlesource.com/boringssl/+/HEAD/crypto/mlkem/mlkem.cc to C code
- Replace all boringssl API calls with OpenSSL API calls, ensuring the [C API is maintained](https://boringssl.googlesource.com/boringssl/+/HEAD/include/openssl/mlkem.h) 
- Use the C API in https://boringssl.googlesource.com/boringssl/+/HEAD/include/openssl/mlkem.h for the implementation of the ML-KEM-768 logic

**Comment by baentsch** [comment]
@nhorman How is a specific "due date" assigned to an issue? Given how central this is, this one imo must be complete by mid November at latest to allow the rest to build on it (and get testing set up around it) so we have enough confidence in the reliability of it by mid February.

FWIW, I now instrumented all MLKEM code in the way I felt necessary but still can't get https://github.com/openssl/openssl/pull/25403 to reliably work OK (1st and 5th and all subsequent handshakes OK; 2nd,3rd,4th always fail). I also went through the KM+KEM code with a fine comb too, but to no avail. I'm now thinking it may have to do with the way the reference code is integrated and would like to switch implementation (to boringssl's as per this issue). Does anyone in the team have experience with both boringssl APIs and OpenSSL APIs, i.e., mapping the former to the latter? The C++->C conversion seems like a second step (for starters we could use `g++` to compile the boringssl code to make use of it).

**Comment by nhorman** [comment]
Normally we assign due dates at the release granularity (you can note this by the assigned milestone above).  If you wish to have more fine grained granularity, I would suggest that, in project 3.5 planning board in the above right panel, you click the item drop downs and assign a start and end date in the corresponding fields there.

speaking only for myself, I do not have any experience with the boring apis, but I'm happy to help if you want to give me specific tasks.  I might also suggest enlisting the help of @andrewkdinh or @Sashan on this subject

**Comment by baentsch** [comment]
> I'm happy to help if you want to give me specific tasks. I might also suggest enlisting the help of @andrewkdinh or @Sashan on this subject

That'd be great. Right now I have the (already somewhat modified but for the reasons above not yet compiling) boringssl code sitting in a branch in the baentsch/openssl repo. Is that the right place to then work together? Do I need to give anyone specific permissions to do so?

**Comment by nhorman** [comment]
Its a fine location for it, though, as a general rule, we tend to use githubs fork functionality to store works in progress (as a matter of branch hygine).  I'm working on the provider tracing feature right now as a POC to present to you to see if its what you need, after which i'll start taking a look at this

**Comment by andrewkdinh** [comment]
@baentsch I have some free cycles this sprint, so feel free to reach out for help with this. I don't have much experience with a similar task, but I’m definitely willing to learn.

**Comment by baentsch** [comment]
@andrewkdinh Thanks very much for that offer. I put up what's available into https://github.openssl.org/mbaentsch/openssl/tree/bsslmlkem768 : Would you please take a look whether you can access that? A `./config && make` in that branch will bring up the (many) problems immediately.

Same question to @nhorman -- plus the question whether I did understand/set up things right with this such as to enable easy sharing within the whole OpenSSL team (but not --yet-- beyond)?

**Comment by baentsch** [comment]
> I now instrumented all MLKEM code in the way I felt necessary but still can't get https://github.com/openssl/openssl/pull/25403 to reliably work OK

Update: This statement no longer holds: MLKEM768 now works reliably OK using the pqcrystals reference code (code in https://github.com/baentsch/openssl/tree/mb-mlkem768). But we cannot use that due to missing CLAs. So importing the boringssl mlkem768 code as replacement is key to completing this task.

**Comment by baentsch** [comment]
fyi, I hit a snag merging the ported boringssl API into the openssl provider logic:

In short, it seems impossible to operate encap on an encoded public key. See https://github.openssl.org/mbaentsch/openssl/blob/bsslmlkem768/include/crypto/mlkem.h: All operations are possible if a fully formed _private_ key is available. From a logical perspective during TLS KEM operations, no problem but the dup/match provider logic then has to work on the (labelled "unstable") internal public & secret key formats _and_ maintain the (specified) encoded public key in parallel (for get/set params and future encode/decode. But the more I think about it, that will be hard-to-impossible to implement using this API, if there's no way to re-create a(n operational) public key from its encoded form: It really seems like boringssl does not _want_ to persist MLKEM (not unlogical, but a challenge if OpenSSL wants to do it -- see discussion we had in Brno @t-j-h @t8m ). Thoughts welcome (as to whether I should simply continue to use/maintain/"transport through the provider interfaces" the internal and encoded key formats in parallel -- or whether someone sees better ways to do this).

**Comment by t-j-h** [comment]
> In short, it seems impossible to operate encap on an encoded public key. 

We simply have to change the boringSSL imported code to make this possible - we aren't at all restricted to maintaining their API approach or chosen architectural limitations.

**Comment by baentsch** [comment]
> > In short, it seems impossible to operate encap on an encoded public key.
> 
> We simply have to change the boringSSL imported code to make this possible - we aren't at all restricted to maintaining their API approach or chosen architectural limitations.

Doing such changes IMO requires complete understanding of that code. At least I'm far away from that right now :-(

On the positive side: I'll then proceed defining an intermediate API to which any mlkem implementation shall conform to support openssl needs. That then also creates a new opportunity for others to contribute to the effort:
 - mapping current bssl API to required ossl API and/or 
 - changing the bssl code as needed and/or 
 - providing a completely different implementation.

**Comment by baentsch** [comment]
I consider this issue functionally resolved and complete if https://github.openssl.org/mbaentsch/openssl/tree/bsslmlkem768 is merged to the public mlkem feature branch. @mattcaswell @t8m please let me know a) whether you agree and b) how to proceed: Should I simply raise a PR from that branch to the feature branch?

**Comment by mattcaswell** [comment]
> I consider this issue functionally resolved and complete if https://github.openssl.org/mbaentsch/openssl/tree/bsslmlkem768 is merged to the public mlkem feature branch. @mattcaswell @t8m please let me know a) whether you agree

Essentially yes. I think there are some aspects that we have deliberately not addressed yet - right? In which case we should ensure that there are separate issues to track any remaining work. As long as those issues are there, I think this issue could be closed once its merged into the feature branch.

> and b) how to proceed: Should I simply raise a PR from that branch to the feature branch?

Seems reasonable.

**Comment by baentsch** [comment]
> As long as those issues are there

Yes, those issues have been created and put (and sorted in priority order) into the overall task list https://github.com/openssl/project/issues/999

> Seems reasonable.

Please let me know when to do ("publish") this.

**Comment by mattcaswell** [comment]
> Please let me know when to do ("publish") this.

IMO we already have the feature branch, so I have no objection to creating a public PR now.

**Comment by baentsch** [comment]
> IMO we already have the feature branch, so I have no objection to creating a public PR now.

After a rebase and preparing this, I stumbled across this message:

![image](https://github.com/user-attachments/assets/ec643bda-7bfe-4ce1-bf96-9ef106035198)

--> it seems GH doesn't know "mbaentsch" and "baentsch" are the same person. (How) can this be fixed?

Oh, and should the complete commit history be retained or should I squash some things?

**Comment by t8m** [comment]
> --> it seems GH doesn't know "mbaentsch" and "baentsch" are the same person. (How) can this be fixed?

You have to submit the PR in https://github.com/openssl/openssl, not in GHE.


> Oh, and should the complete commit history be retained or should I squash some things?

I would squash everything to a single commit (adding Co-authored-by: if there are multiple authors) unless you really want to retain the history. There will be further fixups done during the public PR review.

**Comment by baentsch** [comment]
> > --> it seems GH doesn't know "mbaentsch" and "baentsch" are the same person. (How) can this be fixed?
> 
> You have to submit the PR in https://github.com/openssl/openssl, not in GHE.

How do I do this? Somehow(?) move the work from GHE to the GH "baentsch" account? Will ask later today in person...

> 
> > Oh, and should the complete commit history be retained or should I squash some things?
> 
> I would squash everything to a single commit (adding Co-authored-by: if there are multiple authors) unless you really want to retain the history. There will be further fixups done during the public PR review.

So now done.

**Comment by baentsch** [comment]
@t8m Code now in public GH and PR can now be done as discussed -- but sensibly only to "master": If doing it against the feature branch, a gazillion commits come along (probably everything merged to master since ml-kem feature branch was created): Is that OK? Or could I first update the feature branch to the current master before doing the PR? Or would you (have to) do that?

**Comment by mattcaswell** [comment]
I will rebase the feature branch shortly.

**Comment by mattcaswell** [comment]
I've rebased it. Although it doesn't seem to be showing up in github.com (the rebase is in the master repo, but it's not been mirrored to github.com).

**Comment by baentsch** [comment]
> I've rebased it. Although it doesn't seem to be showing up in github.com (the rebase is in the master repo, but it's not been mirrored to github.com).

So it seems: My compare-for-PR from baentsch/bsslmlkem768 -> openssl/feature/mlkem still shows 78 commits, only the last of which is the one we want... So now what?

**Comment by t8m** [comment]
> > I've rebased it. Although it doesn't seem to be showing up in github.com (the rebase is in the master repo, but it's not been mirrored to github.com).
> 
> So it seems: My compare-for-PR from baentsch/bsslmlkem768 -> openssl/feature/mlkem still shows 78 commits, only the last of which is the one we want... So now what?

The automation to push the rebase from GHE to GH somehow did not trigger. I've triggered it manually. It should be OK now.
